File: examples/python/test_ransac_and_edge.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import numpy as np
import time
from random import random
import open3d as o3d
import cv2

import misc3d as m3d
from utils import np2o3d
from IPython import embed

vis = o3d.visualization.Visualizer()
vis.create_window("Segmentation", 1920, 1200)

color_img = cv2.imread(
    '/home/yuecideng/WorkSpace/Sources/Misc3D/examples/data/indoor/color/color_0.png'
)
color_img = cv2.cvtColor(color_img, cv2.COLOR_BGR2RGB)

depth_img = cv2.imread(
    '/home/yuecideng/WorkSpace/Sources/Misc3D/examples/data/indoor/depth/depth_0.png',
    cv2.IMREAD_ANYDEPTH)

depth = o3d.geometry.Image(depth_img)
color = o3d.geometry.Image(color_img)

rgbd = o3d.geometry.RGBDImage.create_from_color_and_depth(
    color, depth, convert_rgb_to_intensity=False)

pinhole_camera_intrinsic = o3d.camera.PinholeCameraIntrinsic(
    848, 480, 598.7568, 598.7568, 430.3443, 250.244)
pcd = o3d.geometry.PointCloud.create_from_rgbd_image(
    rgbd, pinhole_camera_intrinsic, project_valid_depth_only=True)

print('Point size before sampling', pcd)
pcd = pcd.voxel_down_sample(0.005)
print('Point size after sampling', pcd)

t0 = time.time()
w, index = m3d.common.fit_plane(pcd, 0.01, 100, enable_parallel=True)
print('Plan fitting time: %.3f' % (time.time() - t0))

plane = pcd.select_by_index(index)

t1 = time.time()
index = m3d.features.detect_edge_points(
    plane, o3d.geometry.KDTreeSearchParamHybrid(0.02, 30))
print('Edges detection time: %.3f' % (time.time() - t1))

edges = plane.select_by_index(index)

# scene points painted with gray
m3d.vis.draw_point_cloud(vis, pcd, color=(0.5, 0.5, 0.5))
m3d.vis.draw_point_cloud(vis, plane)
m3d.vis.draw_point_cloud(vis, edges, color=(1, 0, 0), size=5)

vis.run()

File: examples/cpp/test_ransac_and_edge.cpp
#include <iostream>
#include <memory>

#include <misc3d/common/ransac.h>
#include <misc3d/features/edge_detection.h>
#include <misc3d/utils.h>
#include <misc3d/vis/vis_utils.h>
#include <open3d/camera/PinholeCameraIntrinsic.h>
#include <open3d/geometry/Image.h>
#include <open3d/geometry/KDTreeSearchParam.h>
#include <open3d/geometry/PointCloud.h>
#include <open3d/geometry/RGBDImage.h>
#include <open3d/io/ImageIO.h>

int main(int argc, char *argv[]) {
    bool ret;
    open3d::geometry::Image depth, color;
    ret = open3d::io::ReadImage("../examples/data/indoor/depth/depth_0.png",
                                depth);
    ret = open3d::io::ReadImage("../examples/data/indoor/color/color_0.png",
                                color);

    auto rgbd = open3d::geometry::RGBDImage::CreateFromColorAndDepth(
        color, depth, 1000, 3.0, false);

    open3d::camera::PinholeCameraIntrinsic intrinsic(
        848, 480, 598.7568, 598.7568, 430.3443, 250.244);

    auto pcd = open3d::geometry::PointCloud::CreateFromRGBDImage(
        *rgbd, intrinsic, Eigen::Matrix4d::Identity(), false);

    auto pcd_down = pcd->VoxelDownSample(0.005);

    misc3d::Timer timer;
    timer.Start();
    misc3d::common::RANSACPlane fit;
    fit.SetMaxIteration(400);
    fit.SetProbability(0.99);
    fit.SetParallel(true);
    misc3d::common::Plane plane;
    std::vector<size_t> inliers;
    fit.SetPointCloud(*pcd_down);
    ret = fit.FitModel(0.01, plane, inliers);
    std::cout << "Time cost for fitting: " << timer.Stop() << std::endl;

    auto pcd_plane = pcd_down->SelectByIndex(inliers);
    timer.Start();
    auto indices = misc3d::features::DetectEdgePoints(
        *pcd_plane, open3d::geometry::KDTreeSearchParamHybrid(0.02, 30));
    std::cout << "Time cost for edge detection: " << timer.Stop() << std::endl;

    auto pcd_edges = pcd_plane->SelectByIndex(indices);

    auto vis = std::make_shared<open3d::visualization::Visualizer>();
    vis->CreateVisualizerWindow("Ransac and edge detection", 1920, 1200);
    misc3d::vis::DrawPointCloud(vis, *pcd_down, {0.5, 0.5, 0.5});
    misc3d::vis::DrawPointCloud(vis, *pcd_plane);
    misc3d::vis::DrawPointCloud(vis, *pcd_edges, {1, 0, 0});
    vis->Run();

    return 0;
}